### DISCUSSION TITLE
Getting a list of deviceId ?

### DISCUSSION THREAD

**Discussion by R-Nabil** [immich-app/immich#12412] [General]
Hello,

Long story short, i'm trying to re-upload a mess of old pictures that I have. I suspect most of them are already in Immich, but must have been uploaded from a different device.

When I run immich-cli, or immich-go, most of them would be reuploaded (like 99% of them). It might be correct and expected, but I doubt it.

My understanding is that deduplication is also based on the device Id, and Immich-go offers the possibility to override it. So is it possible to get a list of all device Ids that were used to upload pictures in the past in Immich ? If I test with each of them and still get the same upload rate, then I would be confident i'm not going to mess up my setup by uploading all these old pictures.

Thanks

**Comment by danieldietzler** [comment]
Immich automatically de-duplicates your (identical) assets when uploading assets through the api (so either immich-cli, immich-go, mobile, ...). So this shouldn't be a problem at all for you actually, you won't end up with duplicates in immich.

**Reply by R-Nabil** [reply]
Yes I know. But it's a massive amount of extra pictures. They might be from relatives or they might be from me. Going through them will takes ages so I try to use an automatic way first by using this approach. 

If I upload them and realise they are unwanted then I would have screwed my current setup.

**Reply by danieldietzler** [reply]
I am not sure I follow. In your original post you're saying
> My understanding is that deduplication is also based on the device Id

which is incorrect. Immich deduplicates based on identical file content, not metadata.

**Comment by R-Nabil** [comment]
On the server, yes. At the upload level, the device id is involved.

Uploading the same picture exactly from 2 devices would result in 2 uploads. Yes these would be deduplicated later on the server but not at upload level

What I’m trying to simulate here is an upload level de duplication by overriding the deviceId. But we are getting quite away from the original questions that is whether it is possible to get a list of deviceId or not. I’ve seen few suggestions around but none seemed to work.

**Reply by bo0tzz** [reply]
>  Yes these would be deduplicated later on the server but not at upload level

This is incorrect, if you're trying to upload a picture that already exists then it will get rejected immediately by the server, before actually uploading/storing it.

**Reply by R-Nabil** [reply]
> >  Yes these would be deduplicated later on the server but not at upload level
> 
> 
> 
> This is incorrect, if you're trying to upload a picture that already exists then it will get rejected immediately by the server, before actually uploading/storing it. 



Did I under this comment incorrectly ?

https://www.reddit.com/r/immich/s/rmAzk2TXhV

It seems to suggests that when the deviceId for the iPhone has changed everything was reuploaded. 

Or maybe the difference between you and me is a small detail : you mean that the asset would be uploaded then immediately rejected server side so not saved ?

**Reply by bo0tzz** [reply]
The app locally tracks what it has uploaded so far. If the app's data gets cleared that tracking is lost and it will try to upload everything again, but the server will reject those attempts.

**Reply by R-Nabil** [reply]
> The app locally tracks what it has uploaded so far. If the app's data gets cleared that tracking is lost and it will try to upload everything again, but the server will reject those attempts.

Got you. I assumed the asset was uploaded then rejected there. What you're saying is that there is 2 different process 

1. The device keeps track of what has been uploaded
2. Before any upload, a checksum (I guess universal as in any device would compute the same checksum) is exchanged with the server which would reject or authorise the upload. 

By curiosity why are the checksum/id used in 1 and 2 not simply the same ? Is it because computing the one used server side is more costly and that approach improves performance ?

In my case, it means the original question becomes irrelevant. 

Thank you !

**Reply by bo0tzz** [reply]
> why are the checksum/id used in 1 and 2 not simply the same 

Because we haven't gotten around to implementing that yet ;)

**Reply by R-Nabil** [reply]
> > why are the checksum/id used in 1 and 2 not simply the same 
> 
> 
> 
> Because we haven't gotten around to implementing that yet ;)
> 
> 

Yeah makes sense again. 
Thank you for taking the time in answering me.

**Reply by R-Nabil** [reply]
> > why are the checksum/id used in 1 and 2 not simply the same 
> 
> 
> 
> Because we haven't gotten around to implementing that yet ;)
> 
> 

Actually last question : Is there any documentation on what kind of hash is used server side ? Any way this could be computed by myself with some programming involved of course

**Reply by bo0tzz** [reply]
We use an sha1 hash of the file contents.

**Reply by ReanimationXP** [reply]
has this been addressed or changed at all? are the comments above still true? what exactly is happening - what checks are performed when a manual upload is performed? (from the Android client in my case, which I fully cleared in trying to remedy this "issue")

My server and client are both fully up to date with public releases as of 3/16/26. Android 2.5.6 Build 3037. Server 2.5.6.

**Comment by R-Nabil** [comment]
Amazing. Thanks again

**Comment by R-Nabil** [comment]
No. I believe the whole uploading is now checksumming on device. So no duplicate reupload, and I think it is also device independent.

Meaning if you change device you're fine. It will just take a little bit of time at first launch to recompute the hash.

The above is only my understanding

**Comment by aaronbutler1515-cyber** [comment]
Let me know if u need more help here
